File: drills/elixir/harnesses/nc142_lowest_common_ancestor_bst.exs
# Trees are written out directly: nil for empty, {value, left, right} for a
# node. `bud` is a node with no children.
bud = fn value -> {value, nil, nil} end

bst = {6, {2, bud.(0), {4, bud.(3), bud.(5)}}, {8, bud.(7), bud.(9)}}

[
  {"lowest_common_ancestor(bst, 2, 8)", inspect(6),
   inspect(Solution.lowest_common_ancestor(bst, 2, 8))},
  {"lowest_common_ancestor(bst, 2, 4) -- an ancestor counts", inspect(2),
   inspect(Solution.lowest_common_ancestor(bst, 2, 4))},
  {"lowest_common_ancestor(bst, 3, 5)", inspect(4),
   inspect(Solution.lowest_common_ancestor(bst, 3, 5))},
  {"lowest_common_ancestor(bst, 7, 9)", inspect(8),
   inspect(Solution.lowest_common_ancestor(bst, 7, 9))},
  {"lowest_common_ancestor(a single node, 1, 1)", inspect(1),
   inspect(Solution.lowest_common_ancestor(bud.(1), 1, 1))}
]

File: drills/go/harnesses/nc142_lowest_common_ancestor_bst.go
package main

func ancestor(root *TreeNode, p, q int) int {
	return lowestCommonAncestor(root, find(root, p), find(root, q)).Val
}

func main() {
	run(func() []testCase {
		bst := tree(6, 2, 8, 0, 4, 7, 9, x, x, 3, 5)
		return []testCase{
			tc("lowestCommonAncestor(bst, 2, 8)", 6, ancestor(bst, 2, 8)),
			tc("lowestCommonAncestor(bst, 2, 4) -- an ancestor counts", 2, ancestor(bst, 2, 4)),
			tc("lowestCommonAncestor(bst, 3, 5)", 4, ancestor(bst, 3, 5)),
			tc("lowestCommonAncestor(bst, 7, 9)", 8, ancestor(bst, 7, 9)),
			tc("lowestCommonAncestor([1], 1, 1)", 1, ancestor(tree(1), 1, 1)),
		}
	})
}
